pyrobosim/pyrobosim/data/pddlstream/domains/05_nav_grasp_stream/domain.pddl

```diff
@@ -0,0 +1,242 @@
+                       (At ?r ?l1) (not (At ?r ?l2))
+                       (AtPose ?r ?p1)
+                       (NavPose ?l2 ?p2)
+                       (Path ?pth) (Motion ?p1 ?p2 ?pth)
+                  )
+    :effect (and (not (CanMove ?r))
+                 (At ?r ?l2) (not (At ?r ?l1))
+                 (AtPose ?r ?p2) (not (AtPose ?r ?p1))
+                 (increase (total-cost) (PathLength ?pth)))
+  )
+
+  ; PICK: Picks up an object from a specified location
+  (:action pick
+    :parameters (?r ?o ?l ?p ?pr ?g)
+    :precondition (and (Robot ?r)
+                       (Obj ?o)
+                       (Location ?l)
+                       (Pose ?p) (AtPose ?o ?p)
+                       (Pose ?pr) (AtPose ?r ?pr)
+                       (Grasp ?g)
+                       (not (Room ?l))
+                       (HandEmpty ?r) 
+                       (At ?r ?l)
+                       (At ?o ?l)
+                       (Graspable ?o ?p ?pr ?g))
+    :effect (and (Holding ?r ?o) (CanMove ?r)
+                 (not (HandEmpty ?r)) 
+                 (not (At ?o ?l))
+                 (not (AtPose ?o ?p))
+                 (increase (total-cost) (GraspAtPoseCost ?g ?pr)))
+  )
+
+  ; PLACE: Places an object in a specified location
+  (:action place
+    :parameters (?r ?o ?l ?p ?pr)
+    :precondition (and (Robot ?r)
+                       (Obj ?o)
+                       (Location ?l)
+                       (Pose ?p) 
+                       (Pose ?pr) (AtPose ?r ?pr)
+                       (not (Room ?l))
+                       (At ?r ?l)
+                       (not (HandEmpty ?r)) 
+                       (Holding ?r ?o)
+                       (Placeable ?l ?o ?p)
+                       (IsCollisionFree ?l ?o ?p))
+    :effect (and (HandEmpty ?r) (CanMove ?r)
+                 (At ?o ?l) (AtPose ?o ?p)
+                 (not (Holding ?r ?o))
+                 (increase (total-cost) (PickPlaceAtPoseCost ?l ?o ?p ?pr)))
+  )
+
+
+  ; DERIVED PREDICATES
+  ; ISCOLLISIONFREE: Checks that an object pose at a particular location is
+  ;                  collision free with all other objects at that location.
+  (:derived (IsCollisionFree ?l ?o ?p)
+    (not (exists (?obs ?pobs)
+                 (and (Obj ?obs) (Pose ?pobs)
+                      (At ?obs ?l) (AtPose ?obs ?pobs)
+                      (not (CollisionFree ?o ?p ?obs ?pobs)) 
+                 )
+         )
+    ) 
+  )
+  
+  ; HAS: Checks locations using entity and location types or instances, 
+  ;      or even room names as locations
+  (:derived (Has ?loc ?entity)
+    (or 
+        ; CASE 1: Location and entity specified as instances
+        (and (or (Obj ?entity) (Robot ?entity))
+             (Location ?loc)
+             (or (At ?entity ?loc)
+                 (exists (?s) 
+                    (and (Room ?loc) (Location ?s)
+                         (At ?entity ?s) (AtRoom ?s ?loc))
+                 )
+             )
+        )
```

Comment:
?s isn't defined above. Is it a sampled location?

Also, is this reading of the logic correct:

- entity is an object type OR robot type
- AND ?loc is a location type
- entity is at ?loc OR 
  - ?s exists AND:
    - ?loc is a room
    - AND ?s is a location
    - AND ?entity is at ?s 
    - AND ?s is at room ?loc
    
Pretty complicated! In other words, the sampled location and ?loc are the same and ?entity is there?

Reply:
The `exists` predicate is kind of like a lambda, in that you define new variables that only exist in that scope.

And yep, your reading of that logic is exactly right. Those `Has` and `HasAll` derived predicates are pretty tough to get working, which is why PDDL domain definition is kind of an art and people want to get away from it.